File: src-backend/app/models/expression.py
from typing import Any

from pydantic import BaseModel


class ExpressionNode(BaseModel):
    type: str
    value: str
    children: list["ExpressionNode"] = []


class ExpressionResponse(BaseModel):
    expr_id: str
    model_id: str
    latex: str
    complexity: int
    r2_score: float
    tree: ExpressionNode


class ExpressionHistoryEntry(BaseModel):
    operation: str
    latex: str
    complexity: int
    r2_score: float


class ExpressionHistoryResponse(BaseModel):
    expr_id: str
    history: list[ExpressionHistoryEntry]
    current_index: int


class TaskStatusResponse(BaseModel):
    task_id: str
    status: str
    result: dict[str, Any] | None = None
    error: str | None = None

File: src-frontend/src/types/expression.ts
export interface ExpressionNode {
  type: "operator" | "function" | "variable" | "constant";
  value: string;
  children: ExpressionNode[];
}

export interface ExpressionResponse {
  expr_id: string;
  model_id: string;
  latex: string;
  complexity: number;
  r2_score: number;
  tree: ExpressionNode;
  pareto_count: number;
  pareto_index: number;
  variable_impact: Record<string, number>;
  indicators: Record<string, number>;
  target_name: string;
}

export interface ExpressionHistoryEntry {
  operation: string;
  latex: string;
  complexity: number;
  r2_score: number;
}

export interface ExpressionHistoryResponse {
  expr_id: string;
  history: ExpressionHistoryEntry[];
  current_index: number;
}

export interface TaskStatusResponse {
  task_id: string;
  status: "pending" | "running" | "completed" | "failed";
  result: ExpressionResponse | null;
  error: string | null;
}
